samples/bluetooth/encrypted_advertising/central/src/central_ead.c

```diff
@@ -0,0 +1,534 @@
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/bluetooth.h>
+
+#include "common/bt_str.h"
+
+#include "common.h"
+
+LOG_MODULE_REGISTER(ead_central_sample, CONFIG_BT_EAD_LOG_LEVEL);
+
+static uint8_t *received_data;
+static struct key_material keymat;
+
+static bt_addr_le_t peer_addr;
+static struct bt_conn *default_conn;
+
+static struct bt_conn_cb central_cb;
+static struct bt_conn_auth_cb central_auth_cb;
+
+static struct k_poll_signal conn_signal;
+static struct k_poll_signal paired_signal;
+static struct k_poll_signal sec_update_signal;
+static struct k_poll_signal device_found_cb_completed;
+
+/* GATT Discover data */
+static uint8_t gatt_disc_err;
+static uint16_t gatt_disc_end_handle;
+static uint16_t gatt_disc_start_handle;
+static struct k_poll_signal gatt_disc_signal;
+
+/* GATT Read data */
+static uint8_t gatt_read_err;
+static uint8_t *gatt_read_res;
+static uint16_t gatt_read_len;
+static uint16_t gatt_read_handle;
+static struct k_poll_signal gatt_read_signal;
+
+static bool data_parse_cb(struct bt_data *data, void *user_data)
+{
+	if (data->type == BT_DATA_ENCRYPTED_AD_DATA) {
+		int err;
+		struct net_buf_simple decrypted_buf;
+		size_t decrypted_data_size = BT_EAD_DECRYPTED_PAYLOAD_SIZE(data->data_len);
+		uint8_t decrypted_data[decrypted_data_size];
+
+		net_buf_simple_init_with_data(&decrypted_buf, &decrypted_data, decrypted_data_size);
+
+		err = bt_ead_decrypt(keymat.session_key, keymat.iv, data->data, data->data_len,
+				     decrypted_buf.data);
+		if (err < 0) {
+			LOG_ERR("Error during decryption.");
+		}
+
+		bt_data_parse(&decrypted_buf, &data_parse_cb, user_data);
+	} else {
+		LOG_INF("len : %d", data->data_len);
+		LOG_INF("type: 0x%02x", data->type);
+		LOG_HEXDUMP_INF(data->data, data->data_len, "data:");
+
+		/* Copy the data out if we are running in a test context */
+		if (received_data != NULL) {
+			size_t offset = *((size_t *)user_data);
+
+			offset += bt_data_serialize(data, &received_data[offset]);
+
+			*((size_t *)user_data) = offset;
+		} else {
+			*((size_t *)user_data) += bt_data_get_len(data, 1);
+		}
+	}
+
+	return true;
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	int err;
+	size_t parsed_data_size;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	if (default_conn) {
+		return;
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	/* We are only interested in the previously connected device. */
+	if (!bt_addr_le_eq(addr, &peer_addr)) {
+		return;
+	}
+
+	LOG_DBG("Peer found.");
+
+	parsed_data_size = 0;
+	LOG_INF("Received data size: %zu", ad->len);
+	bt_data_parse(ad, data_parse_cb, &parsed_data_size);
+
+	LOG_DBG("All data parsed. (total size: %zu)", parsed_data_size);
+
+	err = bt_le_scan_stop();
+	if (err) {
+		LOG_DBG("Failed to stop scanner (err %d)", err);
+		return;
+	}
+
+	k_poll_signal_raise(&device_found_cb_completed, 0);
```

Comment:
Most other samples use `k_sem` instead of `k_poll_signal`. 

Is there any advantages to `k_poll_signal` over `k_sem`? Otherwise I'd suggest to make it more similar to the other samples we have

Reply:
From what I understand `k_poll` is lighter. And I think here we should stick with `k_poll` because it fit the suggested use of the Zephyr documentation:
[Poll signal suggested uses](https://docs.zephyrproject.org/latest/kernel/services/polling.html#suggested-uses)
[Semaphore suggested uses](https://docs.zephyrproject.org/latest/kernel/services/synchronization/semaphores.html#suggested-uses)

Reply:
That makes sense :)

From the above link:

> Use a poll signal as a lightweight binary semaphore if only one thread pends on it.

Would that indicate that if you use both the system workqueue and a the BT RX thread, you shouldn't use `k_poll`?

Reply:
> Would that indicate that if you use both the system workqueue and a the BT RX thread, you shouldn't use `k_poll` in the main/application thread?

I don't know. @alwa-nordic do you have a thought on that?

Reply:
> > Would that indicate that if you use both the system workqueue and a the BT RX thread, you shouldn't use `k_poll` in the main/application thread?
> 
> I don't know. @alwa-nordic do you have a thought on that?

I'm not sure I understand what the concern is. I don't see any problem.

Reply:
@Thalley I think only the main thread is polling, the only thing that originates from BT_RX and WQ are signals, so it should be fine.